File: app/models/location.rb
class Location < ApplicationRecord
  belongs_to :user
  has_many :reservations
  has_many_attached :images, dependent: :purge

  validates :street, :category, presence: true
  validates :city, presence: true, format: { with: /\A[a-zA-Z\s.-]+\z/ }
  validates :state, presence: true, format: { with: /\A[A-Z][a-z]+(?: +[A-Z][a-z]+)*\z/ }
  validates :country, presence: true
  validates :zip_code, presence: true, format: { with: /\A\d{5}(?:-\d{4})?\z/ }
  validates :notes, presence: true, length: { minimum: 300, maximum: 1000 }
  validates :price, presence: true, format: { with: /\A\d+(?:\.\d{1,2})?\z/ }
end

File: data/Location.js
// models/Location.js

import mongoose from 'mongoose';

const { Schema, model } = mongoose;

const locationSchema = new Schema(
  {
    user: {
      type: mongoose.Schema.Types.ObjectId,
      ref: 'User',
    },
    reservations: [
      {
        type: mongoose.Schema.Types.ObjectId,
        ref: 'Reservation',
      },
    ],
    street: {
      type: String,
      required: true,
    },
    city: {
      type: String,
      required: true,
    },
    state: {
      type: String,
      required: true,
    },
    country: {
      type: String,
      required: true,
    },
    zipCode: {
      type: String,
      required: true,
    },
    price: {
      type: String,
      required: true,
    },
    images: {
      type: [String],
      validate: {
        validator: function (arr) {
          return arr.length == 5;
        },
      },
    },
    notes: {
      type: String,
      required: true,
      minlength: 100,
      maxlength: 600,
    },
    availableDates: [
      {
        startDate: { type: String },
        endDate: { type: String },
      },
    ],
  },
  { timestamps: true }
);

const Location = model('Location', locationSchema);

export default Location;
